From 75bff90b43ad0711efb8c854a8e189771c84f041 Mon Sep 17 00:00:00 2001
From: Thiraviyam Mariyappan <quic_tmariyap@quicinc.com>
Date: Tue, 6 Jun 2023 22:49:13 +0530
Subject: [PATCH] ath12k: Update the number of tx_desc used to FW

With 128 client test,observed low throughput in UDP DL. This is because max
msdu allocated size set as default size(16k) by the FW. So,packet aggregated
in lower rate.
Fixing this issue by setting max msdu allocated size as num of desc size(32k)
through wmi cmd. As a result, Threshold drop reason count reduced and 500mbps
Tput improvement observed from ingress 2400mbps on 5G radio.

Signed-off-by: Thiraviyam Mariyappan <quic_tmariyap@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 7 +++++++
 drivers/net/wireless/ath/ath12k/wmi.h | 1 +
 2 files changed, 8 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -9490,6 +9490,13 @@ static int ath12k_mac_radio_start(struct
 		goto err;
 	}
 
+	ret = ath12k_wmi_pdev_set_param(ar, WMI_PDEV_PARAM_SET_CONG_CTRL_MAX_MSDUS,
+					ATH12K_NUM_POOL_TX_DESC, pdev->pdev_id);
+
+	if (ret) {
+		ath12k_err(ab, "failed to set congestion control MAX MSDUS: %d\n", ret);
+		goto err;
+	}
 	/* Enable(1)/Disable(0) sub channel marking */
 	if (pdev->cap.supported_bands & WMI_HOST_WLAN_5G_CAP) {
 		ret = ath12k_wmi_pdev_set_param(ar, WMI_PDEV_PARAM_SUB_CHANNEL_MARKING,
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -1087,6 +1087,7 @@ enum wmi_tlv_pdev_param {
 	WMI_PDEV_PARAM_RADIO_CHAN_STATS_ENABLE,
 	WMI_PDEV_PARAM_RADIO_DIAGNOSIS_ENABLE,
 	WMI_PDEV_PARAM_MESH_MCAST_ENABLE,
+	WMI_PDEV_PARAM_SET_CONG_CTRL_MAX_MSDUS = 0xa6,
 	WMI_PDEV_PARAM_SUB_CHANNEL_MARKING = 0xb0,
 	WMI_PDEV_PARAM_SET_CMD_OBSS_PD_THRESHOLD = 0xbc,
 	WMI_PDEV_PARAM_SET_CMD_OBSS_PD_PER_AC = 0xbe,
